README.md
# FTP_Server_Teensy41 Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/FTP_Server_Teensy41.svg?)](https://www.ardu-badge.com/FTP_Server_Teensy41)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/FTP_Server_Teensy41.svg)](https://github.com/khoih-prog/FTP_Server_Teensy41/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/FTP_Server_Teensy41.svg)](http://github.com/khoih-prog/FTP_Server_Teensy41/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of contents

* [Table of contents](#table-of-contents)
* [Why do we need this FTP_Server_Teensy41 library](#why-do-we-need-this-FTP_Server_Teensy41-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Teensy boards](#1-for-teensy-boards)
* [Original Documents](#Original-Documents) 
* [Examples](#examples)
  * [ 1. FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2)
  * [ 2. multiFileProject](examples/multiFileProject)
  * [ 3. FTP_Server_SDFAT2_NINA](examples/WiFiNINA/FTP_Server_SDFAT2_NINA) **New**
* [Example FTP_Server_SDFAT2](#example-FTP_Server_SDFAT2)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. FTP_Server_SDFAT2 on Teensy4.1 with QNEthernet](#1-ftp_server_sdfat2-on-teensy-41-with-qnethernet)
  * [2. FTP_Server_SDFAT2 on Teensy4.1 with NativeEthernet](#2-ftp_server_sdfat2-on-teensy-41-with-nativeethernet)
  * [3. FTP_Server_SDFAT2 on TEENSY 4.0 with W5x00 using Ethernet_Generic Library](#3-FTP_Server_SDFAT2-on-TEENSY-40-with-W5x00-using-Ethernet_Generic-Library)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---


### Why do we need this [FTP_Server_Teensy41 library](https://github.com/khoih-prog/FTP_Server_Teensy41)

#### Features

This [**FTP_Server_Teensy41 library**](https://github.com/khoih-prog/FTP_Server_Teensy41) is modified from [Arduino-Ftp-Server library](https://github.com/gallegojm/Arduino-Ftp-Server) to support Teensy 4.x with `QNEthernet`, `NativeEthernet`, `W5x00 using Ethernet_Generic Library` or `Adafruit Airlift Featherwing using WiFiNINA_Generic Library`, and currently SDFat2. In the near future, `any FS, such as PSRAM, LittleFS, (Q)SPI Flash, etc.` will be supported.

---

#### Currently supported Boards

1. **Teensy 4.1 using QNEthernet, NativeEthernet, W5x00 with Ethernet_Generic Library or Adafruit Airlift Featherwing using WiFiNINA_Generic Library** and built-in SDCard, etc.
2. **Teensy 4.0 using W5x00 with Ethernet_Generic Library or Adafruit Airlift Featherwing using WiFiNINA_Generic Library**

---
---

## Prerequisites

 1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
 2. [`Teensy core v1.57+`](https://www.pjrc.com/teensy/td_download.html) for Teensy 4.1
 3. [`QNEthernet Library version v0.16.0+`](https://github.com/ssilverman/QNEthernet) for Teensy 4.1 built-in Ethernet.
 4. [`NativeEthernet Library version stable111+`](https://github.com/vjmuzik/NativeEthernet) for Teensy 4.1 built-in Ethernet.
 5. [`Ethernet_Generic library v2.7.1+`](https://github.com/khoih-prog/Ethernet_Generic) to use **W5x00**. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/Ethernet_Generic.svg?)](https://www.ardu-badge.com/Ethernet_Generic)
 6. [`WiFiNINA_Generic library v1.8.15-1+`](https://github.com/khoih-prog/WiFiNINA_Generic) to use **Adafruit Airlift Featherwing**. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/WiFiNINA_Generic.svg?)](https://www.ardu-badge.com/WiFiNINA_Generic).

---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for `FTP_Server_Teensy41`, then select / install the latest version. You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/FTP_Server_Teensy41.svg?)](https://www.ardu-badge.com/FTP_Server_Teensy41) for more detailed instructions.

### Manual Install

1. Navigate to [FTP_Server_Teensy41](https://github.com/khoih-prog/FTP_Server_Teensy41) page.
2. Download the latest release `FTP_Server_Teensy41-main.zip`.
3. Extract the zip file to `FTP_Server_Teensy41-main` directory 
4. Copy the whole `FTP_Server_Teensy41-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO:

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**FTP_Server_Teensy41** library](https://registry.platformio.org/libraries/khoih-prog/FTP_Server_Teensy41) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/FTP_Server_Teensy41/installation). Search for **FTP_Server_Teensy41** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)

---
---

### Packages' Patches

#### 1. For Teensy boards
 
 **To be able to compile and run on Teensy boards**, you have to copy the files in [**Packages_Patches for Teensy directory**](Packages_Patches/hardware/teensy/avr) into Teensy hardware directory (./arduino-1.8.19/hardware/teensy/avr/boards.txt). 

Supposing the Arduino version is 1.8.19. These files must be copied into the directory:

- `./arduino-1.8.19/hardware/teensy/avr/boards.txt`
- `./arduino-1.8.19/hardware/teensy/avr/cores/teensy/Stream.h`
- `./arduino-1.8.19/hardware/teensy/avr/cores/teensy3/Stream.h`
- `./arduino-1.8.19/hardware/teensy/avr/cores/teensy4/Stream.h`

Whenever a new version is installed, remember to copy this file into the new version directory. For example, new version is x.yy.zz
These files must be copied into the directory:

- `./arduino-x.yy.zz/hardware/teensy/avr/boards.txt`
- `./arduino-x.yy.zz/hardware/teensy/avr/cores/teensy/Stream.h`
- `./arduino-x.yy.zz/hardware/teensy/avr/cores/teensy3/Stream.h`
- `./arduino-x.yy.zz/hardware/teensy/avr/cores/teensy4/Stream.h`

---
---

### Original Documents

Please have a look at [README_orig.md](README_orig.md) or [README.md](https://github.com/gallegojm/Arduino-Ftp-Server/blob/main/README.md)


---
---

### Examples

 1. [FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2)
 2. [multiFileProject](examples/multiFileProject)
 3. [FTP_Server_SDFAT2_NINA](examples/WiFiNINA/FTP_Server_SDFAT2_NINA) **New**

---
---

### Example [FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2)

#### 1. File [FTP_Server_SDFAT2.ino](examples/FTP_Server_SDFAT2/FTP_Server_SDFAT2.ino)

https://github.com/khoih-prog/FTP_Server_Teensy41/blob/edc4cff11ea3c5f5704fd7922a2f04e61bd74d32/examples/FTP_Server_SDFAT2/FTP_Server_SDFAT2.ino#L10-L373


---

#### 2. File [defines.h](examples/FTP_Server_SDFAT2/defines.h)

https://github.com/khoih-prog/FTP_Server_Teensy41/blob/edc4cff11ea3c5f5704fd7922a2f04e61bd74d32/examples/FTP_Server_SDFAT2/defines.h#L10-L126

---
---

### Debug Terminal Output Samples

#### 1. FTP_Server_SDFAT2 on TEENSY 4.1 with QNEthernet

Following is debug terminal output when running example [FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2) on Teensy4.1 using Built-in Ethernet and QNEthernet Library demonstrate the operation of FTP Server and SD Card using SDFat2.


```cpp
Starting FTP_Server_SDFAT2 on TEENSY 4.1 with QNEthernet
FTP_Server_Teensy41 v1.2.0
Initializing SD card...
Wiring is correct and a card is present.

Card type: SDHC

Volume type is FAT32

Volume size (Kbytes): 31452672
Volume size (Mbytes): 30715

===============================
SDCard Initialization : done.
test.txt                                      1048702    22:42  April 29, 2022
foo.txt                                            13    00:00  January 1, 1980
datalog.bin                                  41943040    00:00  January 1, 2022
index.htm                                        3810    22:50  April 29, 2022
mydatalog.txt                                      10    22:24  April 9, 2022
edit.htm.gz                                      4116    20:15  March 13, 2022
CanadaFlag_1.png                                41214    20:15  March 13, 2022
CanadaFlag_2.png                                 8311    20:15  March 13, 2022
CanadaFlag_3.jpg                                11156    20:15  March 13, 2022
favicon.ico                                      1150    22:54  April 29, 2022
graphs.js.gz                                     1971    20:15  March 13, 2022
CanadaFlag_2_1.png                               8311    22:52  April 29, 2022
ESP_AT_WM_Lite.txt                               1547    22:51  April 29, 2022
done!

=========== USE_QN_ETHERNET ===========
Initialize Ethernet using static IP => IP Address = 192.168.2.241
FTP Server Credentials => account = teensy4x, password = ftp_test
[FTP]  Ftp server waiting for connection on port  21
[FTP]  Client connected!
USER teensy4x
PASS ftp_test
[FTP]  Authentication Ok. Waiting for commands.
SYST
DELE test.txt
PORT 192,168,2,30,169,133
[FTP]  Data IP set to  192.168.2.30 , Data port set to  43397
LIST
DELE datalog.bin
PORT 192,168,2,30,135,195
[FTP]  Data IP set to  192.168.2.30 , Data port set to  34755
LIST
PORT 192,168,2,30,165,55
[FTP]  Data IP set to  192.168.2.30 , Data port set to  42295
LIST
PORT 192,168,2,30,206,193
[FTP]  Data IP set to  192.168.2.30 , Data port set to  52929
RETR index.htm
[FTP]  Sending  index.htm
PORT 192,168,2,30,182,29
[FTP]  Data IP set to  192.168.2.30 , Data port set to  46621
STOR index1.htm
[FTP]  Receiving  index1.htm
[FTP]  Transfer completed in (ms)  250 , Speed (kbytes/s)  15
QUIT
[FTP]  Disconnecting client
[FTP]  Ftp server waiting for connection on port  21
```

---


#### 2. FTP_Server_SDFAT2 on TEENSY 4.1 with NativeEthernet

Following is debug terminal output when running example [FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2) on Teensy4.1 using Built-in Ethernet and NativeEthernet Library demonstrate the operation of FTP Server and SD Card using SDFat2.

```cpp
Starting FTP_Server_SDFAT2 on TEENSY 4.1 with NativeEthernet
FTP_Server_Teensy41 v1.2.0
Initializing SD card...
Wiring is correct and a card is present.

Card type: SDHC

Volume type is FAT32

Volume size (Kbytes): 31452672
Volume size (Mbytes): 30715

===============================
SDCard Initialization : done.
index1.htm                                       3810    23:11  April 29, 2022
foo.txt                                            13    00:00  January 1, 1980
index.htm                                        3810    22:50  April 29, 2022
mydatalog.txt                                      10    22:24  April 9, 2022
edit.htm.gz                                      4116    20:15  March 13, 2022
CanadaFlag_1.png                                41214    20:15  March 13, 2022
CanadaFlag_2.png                                 8311    20:15  March 13, 2022
CanadaFlag_3.jpg                                11156    20:15  March 13, 2022
favicon.ico                                      1150    22:54  April 29, 2022
graphs.js.gz                                     1971    20:15  March 13, 2022
CanadaFlag_2_1.png                               8311    22:52  April 29, 2022
ESP_AT_WM_Lite.txt                               1547    22:51  April 29, 2022
done!
======== USE_NATIVE_ETHERNET ========
Using mac index = 7
Connected! IP address: 192.168.2.103
FTP Server Credentials => account = teensy4x, password = ftp_test
[FTP]  Ftp server waiting for connection on port  21
[FTP]  Client connected!
USER teensy4x
PASS ftp_test
[FTP]  Authentication Ok. Waiting for commands.
SYST
PORT 192,168,2,30,146,13
[FTP]  Data IP set to  192.168.2.30 , Data port set to  37389
LIST
DELE index1.htm
PORT 192,168,2,30,170,27
[FTP]  Data IP set to  192.168.2.30 , Data port set to  43547
STOR index1.htm
[FTP]  Receiving  index1.htm
[FTP]  Transfer completed in (ms)  2 , Speed (kbytes/s)  1905
QUIT
[FTP]  Disconnecting client
[FTP]  Ftp server waiting for connection on port  21
```

---


#### 3. FTP_Server_SDFAT2 on TEENSY 4.0 with W5x00 using Ethernet_Generic Library

Following is debug terminal output when running example [FTP_Server_SDFAT2](examples/FTP_Server_SDFAT2) on Teensy4.0 using W5x00 using Ethernet_Generic Library with Large Buffer.

```cpp
Starting FTP_Server_SDFAT2 on TEENSY 4.0 with W5x00 using Ethernet_Generic Library with Large Buffer
FTP_Server_Teensy41 v1.2.0
======== USE_ETHERNET_GENERIC ========
[FTP] Default SPI pinout:
[FTP] MOSI: 11
[FTP] MISO: 12
[FTP] SCK: 13
[FTP] SS: 10
[FTP] =========================
[FTP] Board : TEENSY 4.0 , setCsPin: 10
IP Address = 192.168.2.118
FTP Server Credentials => account = teensy4x, password = ftp_test
[FTP]  Ftp server waiting for connection on port  21
[FTP]  Client connected!
USER teensy4x
PASS ftp_test
[FTP]  Authentication Ok. Waiting for commands.
SYST
PORT 192,168,2,30,233,47
[FTP]  Data IP set to  192.168.2.30 , Data port set to  59695
LIST
PORT 192,168,2,30,148,131
[FTP]  Data IP set to  192.168.2.30 , Data port set to  38019
```

---
---

### Debug

Debug is enabled by default on Serial. Debug Level from 0 to 4. To disable, change the _AWS_TEENSY41_LOGLEVEL_ to 0

```cpp
// Use this to output debug msgs to Serial
#define TEENSY41_DEBUG_PORT                 Serial

// Debug Level from 0 to 4
#define _FTP_SERVER_LOGLEVEL_               2
```

---

### Troubleshooting

If you get compilation errors, more often than not, you may need to install a newer version of Arduino IDE, the Arduino `Teensyduino` core or depending libraries.

Sometimes, the library will only work if you update the `Teensyduino` core to the latest version because I'm always using the latest cores /libraries.


---

### Issues ###

Submit issues to: [FTP_Server_Teensy41 issues](https://github.com/khoih-prog/FTP_Server_Teensy41/issues)

---
---

## TO DO

 1. Fix bug. Add enhancement
 2. Add support to to `LittleFS`, (Q)SPI Flash, etc. besides SD card.


## DONE

 1. Initial porting and coding for **Teensy 4.1 using built-in QNEthernet, NativeEthernet or W5x00 using Ethernet_Generic Library**
 2. Support SD card and `SDFat1`, `SDFat2`
 3. Add debugging features.
 4. Add Table-of-Contents and Version String
 5. Fix bug incomplete downloads from server to client
 6. Add support to `WiFiNINA`, such as `Adafruit Airlift Featherwing`
 7. Configurable `user_name` length to max 63 and `user_password` to max 127


---
---


### Contributions and Thanks

1. Based on and modified from [Jean-Michel Gallego's Arduino-Ftp-Server library](https://github.com/gallegojm/Arduino-Ftp-Server)
2. Thanks to [Warren Watson](https://github.com/wwatson4506) to make PR in [Incomplete downloads from server to client. #2](https://github.com/khoih-prog/FTP_Server_Teensy41/pull/2) leading to the new version `v1.1.0` to fix bug incomplete downloads from server to client
3. Thanks to [davekc](https://github.com/davekc) to post issue in [Setting FTP_CRED_SIZE on the fly like with FTP_BUF_SIZE #4](https://github.com/khoih-prog/FTP_Server_Teensy41/issues/4) leading to the new version `v1.2.0` to have configurable `user_name` length to max 63 and `user_password` to max 127


<table>
  <tr>
    <td align="center"><a href="https://github.com/gallegojm"><img src="https://github.com/gallegojm.png" width="100px;" alt="gallegojm"/><br /><sub><b>⭐️ Jean-Michel Gallego</b></sub></a><br /></td>
    <td align="center"><a href="https://github.com/wwatson4506"><img src="https://github.com/wwatson4506.png" width="100px;" alt="wwatson4506"/><br /><sub><b>Warren Watson</b></sub></a><br /></td>
    <td align="center"><a href="https://github.com/davekc"><img src="https://github.com/davekc.png" width="100px;" alt="davekc"/><br /><sub><b>davekc</b></sub></a><br /></td>
  </tr>
</table>

---

### Contributing

If you want to contribute to this project:

- Report bugs and errors
- Ask for enhancements
- Create issues and pull requests
- Tell other people about this library

---

### License

- The library is licensed under [GPLv3](https://github.com/khoih-prog/FTP_Server_Teensy41/blob/main/LICENSE)

---

## Copyright

- Copyright (c) 2014- Jean-Michel Gallego
- Copyright (c) 2022- Khoi Hoang





changelog.md
# FTP_Server_Teensy41 Library


[![arduino-library-badge](https://www.ardu-badge.com/badge/FTP_Server_Teensy41.svg?)](https://www.ardu-badge.com/FTP_Server_Teensy41)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/FTP_Server_Teensy41.svg)](https://github.com/khoih-prog/FTP_Server_Teensy41/releases)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/FTP_Server_Teensy41.svg)](http://github.com/khoih-prog/FTP_Server_Teensy41/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents


* [Changelog](#changelog)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.0](#releases-v110)
  * [Releases v1.0.0](#releases-v100)

---
---

## Changelog

### Releases v1.2.0

1. Add support to `WiFiNINA`, such as `Adafruit Airlift Featherwing`
2. Configurable `user_name` length to max 63 and `user_password` to max 127. Check [Setting FTP_CRED_SIZE on the fly like with FTP_BUF_SIZE #4](https://github.com/khoih-prog/FTP_Server_Teensy41/issues/4)

### Releases v1.1.0

1. Fix bug incomplete downloads from server to client. Check [Incomplete downloads from server to client. #2](https://github.com/khoih-prog/FTP_Server_Teensy41/pull/2)


### Releases v1.0.0

1. Initial porting and coding for **Teensy 4.1 using built-in QNEthernet, NativeEthernet or W5x00 with Ethernet_Generic Library**
2. Currently supporting only SD Card. Will add support to LittleFS, etc. in next releases



examples/WiFiNINA/FTP_Server_SDFAT2_NINA/defines.h
/****************************************************************************************************************************
  defines.h

  FTP_Server_Teensy41 is an FTP Server for Teensy 4.1 using SD, FS, etc. with QNEthernet or NativeEthernet

  Based on and modified from Arduino-Ftp-Server Library (https://github.com/gallegojm/Arduino-Ftp-Server)
  Built by Khoi Hoang https://github.com/khoih-prog/FTP_Server_Teensy41
 ***************************************************************************************************************************************/

#ifndef defines_h
#define defines_h

#define TEENSY41_DEBUG_PORT                 Serial

// Debug Level from 0 to 4
#define _FTP_SERVER_LOGLEVEL_               1

#if ( defined(CORE_TEENSY) && defined(__IMXRT1062__))
  #if (defined(ARDUINO_TEENSY41))
    // For Teensy 4.1
    #define BOARD_TYPE      "TEENSY 4.1"
    // Use true for QNEthernet or NativeEthernet Library, and false to use other Ethernet_Generic library
    #define USE_QN_ETHERNET         false
    #define USE_NATIVE_ETHERNET     false
    #define USE_WIFI_NINA           true // Added 05-17-22 WEW
  #else
    // For Teensy 4.0
    #define BOARD_TYPE      "TEENSY 4.0"
    // Use false for QNEthernet and NativeEthernet Library, and to use other Ethernet_Generic library
    #define USE_QN_ETHERNET         false
    #define USE_NATIVE_ETHERNET     false
    #define USE_WIFI_NINA           false // Added 05-17-22 WEW
  #endif
#else
  #error Only Teensy 4.1 supported
#endif

#ifndef BOARD_NAME
  #define BOARD_NAME    BOARD_TYPE
#endif

// Use true  for ENC28J60 and UIPEthernet library (https://github.com/UIPEthernet/UIPEthernet)
// Use false for W5x00 and Ethernetx library      (https://www.arduino.cc/en/Reference/Ethernet)

#define USE_UIP_ETHERNET      false
#define USE_ETHERNET_GENERIC  false
#define USE_ETHERNET_ESP8266  false
#define USE_ETHERNET_ENC      false
#define USE_CUSTOM_ETHERNET   false

#if USE_QN_ETHERNET
  #include "QNEthernet.h"       // https://github.com/ssilverman/QNEthernet
  using namespace qindesign::network;
  //#warning Using QNEthernet lib for Teensy 4.1. Must also use Teensy Packages Patch or error
  #define SHIELD_TYPE           "QNEthernet"

#elif USE_NATIVE_ETHERNET
  #include "NativeEthernet.h"
  //#warning Using NativeEthernet lib for Teensy 4.1. Must also use Teensy Packages Patch or error
  #define SHIELD_TYPE           "NativeEthernet"

#elif USE_WIFI_NINA
  //#include "WiFiNINA.h"
  #include "WiFiNINA_Generic.h"
  #include "SPI.h"
  #include "arduino_secrets.h"

  // T4.1 SPI pin defs for WiFiNINA AirLift. To update WiFiNINA_Pinout_Generic.h using this info
  //#warning Using WiFiNINA_Generic lib for Teensy 4.1. Must also update WiFiNINA_Pinout_Generic.h or error

  //#define SPIWIFI       SPI  // The SPI port
  //#define SPIWIFI_SS     5   // Chip select pin
  //#define ESP32_RESETN   6   // Reset pin
  //#define SPIWIFI_ACK    9   // a.k.a BUSY or READY pin
  //#define ESP32_GPIO0   -1
  ///////please enter your sensitive data in the Secret tab/arduino_secrets.h
  char ssid[] = SECRET_SSID;        // your network SSID (name)
  char pass[] = SECRET_PASS;    // your network password (use for WPA, or use as key for WEP)
  #define SHIELD_TYPE           "WiFiNINA"

#elif USE_ETHERNET_GENERIC

  #include <SPI.h>

  ///////////////////////////////////////////////////////////

  // W5100 chips can have up to 4 sockets.  W5200 & W5500 can have up to 8 sockets.
  // Use EthernetLarge feature, Larger buffers, but reduced number of simultaneous connections/sockets (MAX_SOCK_NUM == 2)
  #define ETHERNET_LARGE_BUFFERS

  //////////////////////////////////////////////////////////

  #include "Ethernet_Generic.h"

  #if defined(ETHERNET_LARGE_BUFFERS)
    #define SHIELD_TYPE           "W5x00 using Ethernet_Generic Library with Large Buffer"
  #else
    #define SHIELD_TYPE           "W5x00 using Ethernet_Generic Library"
  #endif

#endif


#if (USE_NATIVE_ETHERNET || USE_ETHERNET_GENERIC)
// Enter a MAC address and IP address for your controller below.
#define NUMBER_OF_MAC      20

byte mac[][NUMBER_OF_MAC] =
{
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x01 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x02 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x03 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x04 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x05 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x06 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x07 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x08 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x09 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x0A },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x0B },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x0C },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x0D },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x0E },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x0F },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x10 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x11 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x12 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0x13 },
  { 0xDE, 0xAD, 0xBE, 0xEF, 0xBE, 0x14 },
};

#else

#define USING_DHCP    true   //true

#if !USING_DHCP
  // Set the static IP address to use if the DHCP fails to assign
  IPAddress myIP(192, 168, 0, 114);
  IPAddress myNetmask(255, 255, 255, 0);
  IPAddress myGW(192, 168, 0, 1);
  //IPAddress mydnsServer(192, 168, 2, 1);
  IPAddress mydnsServer(192, 168, 0, 1);
#endif

#endif

#endif    //defines_h
